test/functional/sqlsrv/sqlsrv_ae_datetimes_as_strings.phpt

```diff
@@ -0,0 +1,627 @@
+    
+    $stmt = sqlsrv_query($conn, "SELECT * FROM [$tableName]");
+    if ($stmt === false) {
+        fatalError("Select from $tableName failed");
+    }
+    
+    while ($row = sqlsrv_fetch_array($stmt, SQLSRV_FETCH_NUMERIC)) {
+        $idnum = $row[0];
+        
+        if ($returnDatesAsStrings == true) {
+            if (!is_string($row[1])) {
+                fatalError("String for date expected, not a string");
+            }
+            
+            CompareDateTime($datetimetype, $returnDatesAsStrings, $expectedDateTime, $row[1]);
+        } else {
+            if (!($row[1] instanceof DateTime)) {
+                fatalError("DateTime object expected, not a DateTime");
+            }
+
+            $datetime = ($datetimetype == 'datetimeoffset') ? date_format($row[1], 'Y-m-d H:i:s.u P') : date_format($row[1], 'Y-m-d H:i:s.u');
+            
+            CompareDateTime($datetimetype, $returnDatesAsStrings, $expectedDateTime, $datetime);
+        }
+    }
+    
+    print_r(sqlsrv_field_metadata($stmt)[1]);
+
+}
+
+// The date used for the test will be Januray 31, 2002, or 2002/01/31.
+// This will sidestep issues involving the use of two digit years since 
+// SQL Server defaults to 19 as the first two digits.
+// Time is 23:59:29.049876
+// Note that smalldatetime rounds to the nearest minute, and that may cause 
+// this test to fail if it rolls over to the next day.
+// Incidentally, this datetime corresponds to a timestamp of 1012521599.
+$year = '2002';
+$month = '01';
+$month_name = 'January';
+$month_abbr = 'Jan';
+$day = '31';
+$hour = '23';
+$hour12 = '11';
+$meridian = 'PM';
+$minute = '59';
+$second = '29';
+$frac = '04';
+$frac2 = '9876';
+$tz_correction = '+08:00';
+
+// This is the array of dates/times/timezones to test against. They have
+// different numbers of trailing zeroes to match the precision of different
+// SQL date and time types, but all go up to microseconds because that's
+// how PHP formats times with date_format(), allowing direct string comparisons
+// when the DateTime objects retrieved from a table are formatted as strings
+// with date_format().
+$expectedDateTime = array($year."-".$month."-".$day." ".$hour.":".$minute.":".$second.".".$frac.$frac2,
+                          $year."-".$month."-".$day." ".$hour.":".$minute.":".$second.".".$frac."0000",
+                          $year."-".$month."-".$day." ".$hour.":".$minute.":".$second.".000000",
+                          $year."-".$month."-".$day." ".$hour.":".$minute.":00.000000",
+                          $year."-".$month."-".$day." ".$hour.":".$minute.":".$second.".".$frac.$frac2." ".$tz_correction,
+                          $year."-".$month."-".$day." ".$hour.":".$minute.":".$second.".".$frac."0000 ".$tz_correction,
+                          $year."-".$month."-".$day." ".$hour.":".$minute.":".$second.".000000 ".$tz_correction,
+                          $year."-".$month."-".$day." ".$hour.":".$minute.":00.000000 ".$tz_correction,
```

Comment:
is this case ever going to occur? time only has minute precision when the type is smalldatetime, however then it's smalldatetime, there should not be offset right?

Reply:
Yes, here I can remove this format.

Reply:
Actually it is needed here too, because we are inserting and retrieving all date/time types. If I remove this, a smalldatetime retrieved using date_format() gives something like 2002-01-31 23:59:00.000000, so this case can occur.

Reply:
Then you can remove this comparison in CompareDateTimeString (line 110)? The retrieve value in this function doesn't use date_format().